tools/fp20compiler/main.cpp

```diff
@@ -34,30 +34,139 @@ nvparse_errors errors;
 int line_number;
 char * myin = 0;
 
-void translate(const char* s) {
-    char* ns = strdup(s);
+static char* copy_string(const char* str, size_t length) {
+    char* buffer = (char*)malloc(length + 1);
+    memcpy(buffer, str, length);
+    buffer[length] = '\0';
+    return buffer;
+}
 
-    // printf("-- %s\b", s);
+static char* find_at_line_start(const char* haystack, const char* cursor,
+                                const char* needle) {
+    cursor = strstr(cursor, needle);
+    while (cursor != NULL) {
 
-    char* ts_start = strstr(ns, "!!TS1.0");
-    if (ts_start) {
-        char* ts_end = strstr(ts_start, "// End of program");
-        assert(ts_end);
-        *ts_end = 0;
+        // Accept if candidate is at beginning of file or line
+        if ((cursor == haystack) || (cursor[-1] == '\n')) {
+            return (char*)cursor;
+        }
 
-        ts10_init(ts_start);
-        ts10_parse();
+        cursor = strstr(cursor + 1, needle);
+    }
+    return NULL;
+}
 
-        *ts_end = '/';
+static char* skip_until(const char* str, const char* delim) {
```

Comment:
I didn't know / remember `strpbrk` probably. When reading the `man`-page I also noticed `strcspn` which is even closer to what I was implementing.

The new revision I pushed is using `strcspn`.

Reply:
I've also replaced `copy_string` by `strndup`.
I assume the original reason for not using `strndup` was to avoid non-C-standard functions?
However the existing code already used `strdup`, too.

Another motivation might have been to avoid counting the length again.
However, this code path isn't performance critical, and there's probably no real difference to begin with.